File: packages/cli/src/commands/logout.ts
import type { Command } from 'commander';
import { icons, c, log, logError, logJson, getCurrentMode } from '../lib/output.js';
import { loadConfig, removeToken } from '../lib/config.js';
import { createApiClient, ApiClientError } from '../lib/api-client.js';

export const registerLogoutCommand = (program: Command): void => {
  program
    .command('logout')
    .description('Remove saved authentication token')
    .action(async () => {
      const config = loadConfig();

      if (!config.token) {
        if (getCurrentMode() === 'json') {
          logJson({ status: 'not_logged_in' });
          return;
        }
        logError(
          'Not logged in.',
          undefined,
          `Run ${c.cmd('spm login')} to authenticate with GitHub.`,
        );
        process.exitCode = 1;
        return;
      }

      // Attempt to invalidate token on server (non-fatal if it fails)
      try {
        const client = createApiClient(config);
        await client.logout();
      } catch (err) {
        if (!(err instanceof ApiClientError)) {
          // Network error is non-fatal for logout
        }
      }

      // Remove token locally
      removeToken();

      if (getCurrentMode() === 'json') {
        logJson({ status: 'logged_out' });
        return;
      }

      console.log('');
      log(`${icons.success} Logged out. Token removed from ${c.path('~/.spm/config.toml')}.`);
      console.log('');
      log(`You can still search and install skills.`);
      log(`Publishing requires ${c.cmd('spm login')}.`);
    });
};

File: cli-go/cmd/logout.go
package cmd

import (
	"fmt"

	"github.com/skillpkg/spm/internal/api"
	"github.com/skillpkg/spm/internal/config"
	"github.com/skillpkg/spm/internal/output"
	"github.com/spf13/cobra"
)

var logoutCmd = &cobra.Command{
	Use:   "logout",
	Short: "Remove saved authentication token",
	Long:  "Clear the saved authentication token from ~/.spm/config.toml and revoke it on the server.",
	RunE:  runLogout,
}

func init() {
	rootCmd.AddCommand(logoutCmd)
}

func runLogout(cmd *cobra.Command, _ []string) error {
	if Cfg.Token == "" {
		if Out.Mode == output.ModeJSON {
			return Out.LogJSON(map[string]string{"status": "not_logged_in"})
		}
		Out.LogError("Not logged in.")
		Out.Log("Run %s to authenticate with GitHub.", output.Cyan("spm login"))
		return fmt.Errorf("not logged in")
	}

	// Attempt to invalidate token on server (non-fatal if it fails)
	client := api.NewClient(Cfg.RegistryURL(), Cfg.Token)
	_ = client.Logout()

	// Remove token locally
	Cfg.Token = ""
	if err := config.Save(Cfg); err != nil {
		Out.LogError("Failed to save config: %s", err)
		return fmt.Errorf("failed to clear token")
	}

	if Out.Mode == output.ModeJSON {
		return Out.LogJSON(map[string]string{"status": "logged_out"})
	}

	Out.Log("")
	Out.Log("%s Logged out. Token removed from %s.", output.Icons["success"], output.Dim("~/.spm/config.toml"))
	Out.Log("")
	Out.Log("You can still search and install skills.")
	Out.Log("Publishing requires %s.", output.Cyan("spm login"))

	return nil
}
